File: observer/java/ObservableProduct.java
package Observer;

import java.util.ArrayList;
import java.util.List;

// Takip edilecek olan yapımızdır.
// İçerisinde Observer (değişimden etkilenecek) nesneleri tutar.
// UML diyagramındaki Observable yapısına denk gelmektedir.
abstract class ObservableProduct {
    private List<IObserverUser> _users;
    protected ProductUpdateMessage message;

    public ObservableProduct() {
        _users = new ArrayList<IObserverUser>();
        message = new ProductUpdateMessage();
    }

    public void addObserver(IObserverUser observerUser) {
        _users.add(observerUser);
    }

    public void removeObserver(IObserverUser observerUser) {
        _users.remove(observerUser);
    }

    public void notifyObserver() {
        for (IObserverUser user : _users) {
            user.sendNotification(message);
        }
    }
}

File: observer/kotlin/ObservableProduct.kt
package Observer

import java.util.ArrayList

// Takip edilecek olan yapımızdır.
// İçerisinde Observer (değişimden etkilenecek) nesneleri tutar.
// UML diyagramındaki Observable yapısına denk gelmektedir.
internal abstract class ObservableProduct {
    private val _users: MutableList<IObserverUser>
    protected var message: ProductUpdateMessage

    init {
        _users = ArrayList()
        message = ProductUpdateMessage()
    }

    fun addObserver(observerUser: IObserverUser) {
        _users.add(observerUser)
    }

    fun removeObserver(observerUser: IObserverUser) {
        _users.remove(observerUser)
    }

    fun notifyObserver() {
        for (user in _users) {
            user.sendNotification(message)
        }
    }
}